host/discoverd.go

```diff
@@ -0,0 +1,111 @@
+package main
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/flynn/flynn/discoverd/client"
+	"github.com/flynn/flynn/host/logmux"
+	"github.com/flynn/flynn/pkg/shutdown"
+)
+
+func NewDiscoverdManager(backend Backend, mux *logmux.LogMux, hostID, publishAddr string) *DiscoverdManager {
+	d := &DiscoverdManager{
+		backend: backend,
+		mux:     mux,
+		inst: &discoverd.Instance{
+			Addr: publishAddr,
+			Meta: map[string]string{"id": hostID},
+		},
+	}
+	shutdown.BeforeExit(d.Close)
+	return d
+}
+
+type DiscoverdManager struct {
+	backend Backend
+	mux     *logmux.LogMux
+	inst    *discoverd.Instance
+	mtx     sync.Mutex
+	local   discoverd.Heartbeater
+	peer    discoverd.Heartbeater
+}
+
+func (d *DiscoverdManager) Close() {
+	d.mtx.Lock()
+	defer d.mtx.Unlock()
+	if d.local != nil {
+		d.local.Close()
+		d.local = nil
+	}
+	if d.peer != nil {
+		d.peer.Close()
+		d.peer = nil
+	}
+}
+
+func (d *DiscoverdManager) localConnected() bool {
+	d.mtx.Lock()
+	defer d.mtx.Unlock()
+	return d.local != nil
+}
+
+func (d *DiscoverdManager) ConnectLocal(url string) error {
+	if d.localConnected() {
+		return errors.New("host: discoverd is already configured")
+	}
+	disc := discoverd.NewClientWithURL(url)
+	// TODO: use attempts to connect
```

Comment:
discoverd should probably use a retry dialer like [the controller client](https://github.com/flynn/flynn/blob/master/controller/client/client.go#L134).